### PR TITLE
Jwt dig

### PR DESCRIPTION
**Pull request by sashaCher**
### What does this PR do?

POC for JWT hierarchy 

#### Change log
- [ ] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [ ] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [ ] This PR does not require updating any documentation

#### API Changes
- [ ] The [OpenAPI spec](https://github.com/cyberark/conjur-openapi-spec) has been updated to meet new API changes (or an issue has been opened), or
- [ ] The changes in this PR do not affect the Conjur API

### CODE REVIEW

File: app/domain/authentication/authn_jwt/restriction_validation/validate_restrictions_one_to_one.rb [hunk 1]
```diff
@@ -22,12 +22,13 @@ def valid_restriction?(restriction)
             raise Errors::Authentication::ResourceRestrictions::EmptyAnnotationGiven, annotation_name
           end
 
-          unless @decoded_token.key?(claim_name)
+          token_path = claim_name.split('/')
+          token_value = @decoded_token.dig(*token_path)
```
Comment (tzheleznyak) [inline app/domain/authentication/authn_jwt/restriction_validation/validate_restrictions_one_to_one.rb:26]: What happens if path not exist in the token payload?

File: app/domain/authentication/authn_jwt/identity_providers/identity_from_decoded_token_provider.rb [hunk 2]
```diff
@@ -63,7 +63,8 @@ def fetch_identity_name_from_token
           return @identity_name_from_token if @identity_name_from_token
 
           @logger.debug(LogMessages::Authentication::AuthnJwt::CheckingIdentityFieldExists.new(token_id_field_secret))
-          @identity_name_from_token = decoded_token[token_id_field_secret]
+          token_path = token_id_field_secret.split('/')
+          @identity_name_from_token = decoded_token.dig(*token_path)
```
Comment (tzheleznyak) [inline app/domain/authentication/authn_jwt/identity_providers/identity_from_decoded_token_provider.rb:67]: Maybe that part fetching a value of internal token claim should be seperated to a class that this class and validate_restrictions_one_to_one both use?

File: app/domain/authentication/authn_jwt/consts.rb [hunk 3]
```diff
@@ -31,7 +31,7 @@ module AuthnJwt
     CACHE_MAX_CONCURRENT_REQUESTS = 3
     MANDATORY_CLAIMS = [EXP_CLAIM_NAME].freeze
     OPTIONAL_CLAIMS = [ISS_CLAIM_NAME, NBF_CLAIM_NAME, IAT_CLAIM_NAME].freeze
-    VALID_CLAIM_NAME_REGEX = /^[a-zA-Z|$|_][a-zA-Z|$|_|0-9|.]*$/.freeze # starts with letter $ or _ can contains digits and dot
+    VALID_CLAIM_NAME_REGEX = /^[a-zA-Z|$|_][a-zA-Z|$|_|0-9|.\/]*$/.freeze # starts with letter $ or _ can contains digits and dot
```
Comment (nessiLahav) [inline app/domain/authentication/authn_jwt/consts.rb:34]: change comment as well

Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/consts.rb:34]: @sashaCher WDYT to leave the previous `VALID_CLAIM_NAME_REGEX` logic and to add a command class which handle a nested claim name?
And this class responsibility will be to split the claim name by the delimiter  / and maybe to throw an error about invalid format

**Review by nessiLahav — CHANGES_REQUESTED** [review]
@benoaviram Just want to raise that we had a bad experience with skipping the regular flow of functional sign off and high level design in the past.

it is different process to review the code and to review the idea.

File: app/domain/authentication/authn_jwt/consts.rb (hunk 3 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/consts.rb]: @nessiLahav See now
I think that creating separated class for nested validation is over-architecture and will make code more complicated.
The responsibility of the class it "claim related variables" input validation rather "single claim section" input validation.
You're welcome to propose a new name for the class reflects the class responsibility.

File: app/domain/authentication/authn_jwt/identity_providers/identity_from_decoded_token_provider.rb (hunk 2 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/identity_providers/identity_from_decoded_token_provider.rb]: @nessiLahav @tzheleznyak I've took split and dig procedure to a separated class

**Pull request closed without merging**